### DISCUSSION TITLE
How do you create type registry JSON files?

### DISCUSSION THREAD

**Discussion by khssnv** [polkascan/py-scale-codec#96] [Q&A]
Hi there :wave:

There are type registry JSON files in a [type_registry](https://github.com/polkascan/py-scale-codec/tree/master/scalecodec/type_registry) directory. It seems like very few of them are provided by projects developers. Could you reveal how you create these JSONs? It could help to contribute to the project. Thanks!

**Comment by arjanz** [comment]
There is some history for those type registry files. Since Substrate introduced `MetadataV14`, it contains an embedded type registry with all types used in the calls and storage functions. Before that, types were just mere strings how they were named in the RUST code of the runtime. So back then there was a need to manually administer a type registry to describe the type decomposition in primitives. 

For networks like Kusama those registries were already maintained by PolkadotJS ([example](https://github.com/polkadot-js/api/blob/69e754105d9d2b6f2eaac750ba5d61082cbae924/packages/types-known/src/spec/kusama.ts)), so I could gratefully re-use those. 

But for other networks that don't provide an own registry file, you basically have to analyse the RUST source and create a type registry manually:

Crust example: https://github.com/crustio/crust/blob/d2de3d018c9367cec98411d56fdb0836cffb75cd/cstrml/benefits/src/lib.rs#L138

Type registry: https://github.com/polkascan/py-scale-codec/blob/6c897a0f09d3fadcb84eae393d98a831cf403b50/scalecodec/type_registry/crust.json#L28

But later on they upgrade to `MetadataV14`, so there was no more need for the manually type registry. (See comment: https://github.com/polkascan/py-substrate-interface/issues/251#issuecomment-1261431740)

TL;DR: Probably you don't need to create/maintain a type registry file, unless the Substrate runtime still utilises `MetadataV13` or lower (check `substrate.get_metadata()` and `substrate.implements_scaleinfo()`).

**Comment by khssnv** [comment]
Got it, thank you for sharing :+1:
